### ISSUE TITLE
XHR's displaying as 'aborted' when they are not

### ISSUE THREAD

**Issue by bdruth** [cypress-io/cypress#1652]
### Current behavior:
When running a cypress spec, many (but not all) XHRs are shown as 'aborted' in the Cypress test runner when they are not, in fact. Chrome dev tools `Network` tab shows successful XHRs and the app is working as designed, completing requests with no errors. We have one spec, currently, and the spec contains three tests: login / access app / use app. During the _access app_ test, we start `cy.server()` and route 2 requests with mock payloads. These are the only requests mocked in the spec. In this test and the preceding test, all XHRs look fine in the runner. In the subsequent test, however, the XHRs all show aborted - they show as red.

#### test 2/3 looks like:
![ok](https://gist.githubusercontent.com/bdruth/d3d1881dbd849774f910cf530fd5ce09/raw/871997f3891a21a175b29e28b461a85533f0cde3/ok.png)

#### test 3/3 looks like:
![aborted](https://gist.githubusercontent.com/bdruth/d3d1881dbd849774f910cf530fd5ce09/raw/871997f3891a21a175b29e28b461a85533f0cde3/aborted.png)

Per [conversation](https://gitter.im/cypress-io/cypress?at=5ae727571388b2df2113e243) with @ValerieThoma on Gitter, I added `cy.wait()` for some routes in the aborted requests. This didn't change the behavior as near as I can tell.

#### test 3/3 with cy.wait looks like:
![aborted with wait](https://gist.githubusercontent.com/bdruth/d3d1881dbd849774f910cf530fd5ce09/raw/871997f3891a21a175b29e28b461a85533f0cde3/xhr_with_wait.png)

Additional info: the app communicates via XHR to multiple backends - probably 3 different hosts, amongst them Amazon S3 - all XHRs show as 'aborted' - so this seems like something that's local to the client app / cypress runner interaction. The app is Angular 5 w/ HttpClient.

### Desired behavior:
Successful XHRs should not display as aborted.

### Steps to reproduce:
Not sure how to reproduce generically, unfortunately.

### Versions  
Cypress 2.1.0
Chrome 66

**Comment by pavankumarkatakam** [comment]
Im also geting this error, cypress UI is showing aborted but in cosole it is showing status ok and in network also it is showing ok and futhuer xhr requests is processing as well. I think only UI needs to be changed.

**Comment by ronlawrence3** [comment]
I also had this symptom which is related to #761 . If you open your browser console and see the error during your test then it is probably related.

```
UnsubscriptionError: 1 errors occurred during unsubscription:
  1) TypeError: Cannot set property 'aborted' of undefined
```

**Comment by pavankumarkatakam** [comment]
@ronlawrence3 , Did you find any fix for this

**Comment by ronlawrence3** [comment]
unfortunately, no. I added a 2 second sleep to my afterEach to work around it, which is NOT ideal

**Comment by nikosdouvlis** [comment]
This also happens with Cypress 3.1.0 sadly. Any ideas?

**Comment by smclt30p** [comment]
In my case, the app was  a single page application, and a form submit causes Cypress to expect a page reload and it cancels the XHR because of that. So I changed my login form from a form to a dumb div, so that cypress does not expect a page reload.

**Comment by pavankumarkatakam** [comment]
I think this is the same problem for me also, i have a form tag in Angular app. But there are many form tags in whole application.

**Comment by smclt30p** [comment]
Are you firing XHR's manually from that form submit?

**Comment by pavankumarkatakam** [comment]
Yes, im using the angular http on submit button click and after doing some logic then im redirecting to profile page (but that is angular internal page rather than full page reload.)

**Comment by smclt30p** [comment]
Cypress cancels the XHR because of that. Change the form to a div.

On 7 Sep 2018, at 10:43, Pavan kumar <notifications@github.com<mailto:notifications@github.com>> wrote:


Yes, im using the angular http on submit button click and after doing some logic then im redirecting to profile page (but that is angular internal page rather than full page reload.)

**Comment by mustafahamid** [comment]
what is the solution for this problem please ?

**Comment by pavankumarkatakam** [comment]
Hi @smclt30p , If i change **form** to **div**, then the angular framework will complain about that div doesn't have **exportAs** feature for **ngForm** Directive. I need to check if there is another solution.

@mustafahamid , Did you find any solution.

**Comment by mlc-mlapis** [comment]
... couldn't be related to same problem as in https://github.com/assaf/zombie/issues/1079 and https://github.com/assaf/zombie/pull/1097 ?

**Comment by mlc-mlapis** [comment]
... it's still a problem in all of our tests ... app XHR requests are showed as aborted ...

![image](https://user-images.githubusercontent.com/5693835/47647406-344f3f00-db77-11e8-84e0-0e299940fadb.png)

but the console shows the opposite:

![image](https://user-images.githubusercontent.com/5693835/47647541-93ad4f00-db77-11e8-9b98-ecbf371c449d.png)

The main problem is the question if that behavior affects the tests processing or it's just an unimportant side effect problem in UI.

**Comment by Olgagr** [comment]
We have the same issue as well. We test an Angular app and all requests are reported as aborted (GET, POST, PUT, DELETE). In the dev console, everything looks ok. Moreover, we can wait for a request to finish (```cy.wait```). 

Cypress 3.1.0
MacOS 10.13.6
Node 8.12.0
Chrome 70.0.3538.77

**Comment by mlc-mlapis** [comment]
@Olgagr ... hi, you mentioned you are testing Angular app. You use the same technology, and we are still trying a way how to control the timing of all async calls which run under the cover when you go through the app. Sure, `.wait()` is useful but I still think that it is necessary to introduce and use something more predictable and controlled. The app itself uses observable / subscribe pattern and each point knows what have to ready to allow to run it.

So we are thinking about some logic and syntax as ... `hey CY, wait till the moment when some point in defined interface against the app API will contain that value ... then continue` ... and not explicitly code `.wait(5000)`.

Now we are discussing what should be such app API against CY and what type of action/command it should be in CY, which would wait and monitor that app API.

What do you think about the theme?

**Comment by Olgagr** [comment]
@mlc-mlapis Hi, right now I'm using ```cy.wait``` with request's aliases eg.:

```
    beforeEach(function() {
      cy.server();
      cy.route('GET', '/admin/api/clients/*').as('getClient');
    });

    it('some test', function () {
        cy.wait('@getClient');
    });
```

It works great, despite the Cypress reporting the request to be aborted which is annoying. This way we avoid iffy method like ```cy.wait(5000)```

Have you tried this approach?

**Comment by smclt30p** [comment]
Are you making your request from buttons that are inside of form elements?

On Nov 2, 2018, at 3:49 PM, Olga Grabek <notifications@github.com<mailto:notifications@github.com>> wrote:


@mlc-mlapis<https://github.com/mlc-mlapis> Hi, right now I'm using cy.wait with request's aliases eg.:

    beforeEach(function() {
      cy.server();
      cy.route('GET', '/admin/api/clients/*').as('getClient');
    });

    it('some test', function () {
        cy.wait('@getClient');
    });


It works great, despite the Cypress reporting the request to be aborted which is annoying. This way we avoid iffy method like cy.wait(5000)

Have you tried this approach?

**Comment by Olgagr** [comment]
@smclt30p This one ('/admin/api/clients/*') is of course not from the form element. However, I have other ones that are from buttons that are inside of form elements. No matter what, all are reported as aborted.

**Comment by mlc-mlapis** [comment]
@Olgagr ... yep, actual syntax and functionality of aliases are super power, but still not the 100% solution for all cases with XHR requests, which run in an app on the background.

Especially I mean cases where it's not certain if such a XHR request will exist or not or it was resolved a test before, because it depends on the timing of the previous test (`it`). I mean cases where the tests are not fully independent because of you are testing some scenarios ... going through some screens, click by click ... so you have several `it` parts and of course one `beforeEach` with `cy.server(); cy.route('GET', globMask).as(getRouteAlias);` ...

Then you can get situations where the XHR request is once called in the test but then in its repeating not or is called in `beforeEach` part or even as a part of the previous test ... so it is not detected in the test itself.

So, I was talking about the possibility to have syntax like:
* `cy.wait('&myAppState')` where the test would wait till the moment when `myAppState` would exist
* `cy.wait('[&myAppState=true]', {timeout: 10000})` where the state exist and would represent the value `true`

**The question is what should be such app states.** The simplest way would be the global `window` / `document` property with the name equal to the name used in `cy.wait('&myAppState')`, so `myAppState`. But it's not probably the only way how to do it.

**Comment by janhjordie** [comment]
I'm also having the same issue. 
And it's only happening in Cypress. 

Using Cypress 3.1.1, Electron 59 and Chrome 70.

I'm trying to listen on GET's from signalr, and wan't to collect data from the poll results.
But since the last GET always aborts, everything stops. Our live site works just fine. The abort only happens in Cypress

![cypress-sandbox](https://user-images.githubusercontent.com/79910/48587735-99a37e00-e934-11e8-94f2-4ece6d30ae2a.jpg)

Please help !

**Comment by tonytytran** [comment]
I am also having the same issue as well, where testing our site manually, calls are being made and returning true as expected, but are being aborted occasionally in cypress.

Please help x2!

**Comment by jennifer-shehane** [comment]
Unfortunately in order to address an issue, we have to have a reproducible example provided. Can anyone provide the exact code + test code to reproduce this?

**Comment by janhjordie** [comment]
@jennifer-shehane could you supply me with a email I can post my code to ?

**Comment by jennifer-shehane** [comment]
@janhjordie support@cypress.io

**Comment by azollai** [comment]
I am getting the same "aborted" message while my requests in cypress. Though there is no problem with it on network panel. I think it is a really important issue to solve...

**Comment by brian-mann** [comment]
These requests are not actually being aborted. The problem is that Angular uses `rxJS` under the hook, and the disposer function calls `xhr.abort()`, which is called whenever an XHR completes.

This calls the Cypress `xhr#abort` method, which then visually sets the XHR as being aborted. It's actually only a **visual** bug, its not a functional bug. The comments related to things other than visual are not actually relevant to this particular issue and should be opened as separate issues. I'm fixing this now.

**Comment by brian-mann** [comment]
I've [opened a separate issue here](https://github.com/cypress-io/cypress/issues/2968) for not manually aborting XHR's aggressively on unload.

This should fix the rest of the issues in this thread. Once this PR is merged in I will lock this issue, and additional comments or problems should go in a new issue, as I believe these are all addressed.

**Comment by jennifer-shehane** [comment]
The code for this is done, but this has yet to be released. We'll update this issue and reference the [changelog](https://on.cypress.io/changelog) when it's released.

**Comment by brian-mann** [comment]
Released in [`3.1.4`](https://docs.cypress.io/guides/references/changelog.html#3-1-4).
